src/validator.cpp

```diff
@@ -1355,4 +1389,178 @@ void Validator::ValidatorImpl::checkUnitForCycles(const ModelPtr &model, const U
+        if (!units->id().empty()) {
+            if (units->isImport()) {
+                info = "  - imported units '" + units->name() + "' in model '" + model->name() + "'\n";
+            } else {
+                info = "  - units '" + units->name() + "' in model '" + model->name() + "'\n";
+            }
+            addIdMapItem(units->id(), info, idMap);
+        }
+        for (size_t i = 0; i < units->unitCount(); ++i) {
+            std::string reference;
+            std::string prefix;
+            double exponent;
+            double multiplier;
+            std::string id;
+            units->unitAttributes(i, reference, prefix, exponent, multiplier, id);
+            if (!id.empty()) {
+                info = "  - unit in units '" + units->name() + "' in model '" + model->name() + "'\n";
+                addIdMapItem(id, info, idMap);
+            }
+        }
+        if (units->isImport() && units->importSource() != nullptr && !units->importSource()->id().empty()) {
+            info = "  - import source for units '" + units->name() + "'\n";
+            addIdMapItem(units->importSource()->id(), info, idMap);
+        }
+    }
+    // Encapsulation.
+    if (!model->encapsulationId().empty()) {
+        info = "  - encapsulation in model '" + model->name() + "'\n";
+        addIdMapItem(model->encapsulationId(), info, idMap);
+    }
+
+    // Start recursion through encapsulation hierarchy.
+    for (size_t c = 0; c < model->componentCount(); ++c) {
+        buildComponentIdMap(model->component(c), idMap, reportedConnections);
+    }
+    return idMap;
+}
+
+void Validator::ValidatorImpl::buildComponentIdMap(const ComponentPtr &component, IdMap &idMap, std::set<std::string> &reportedConnections)
+{
+    std::string info;
+
+    // Component.
+    if (!component->id().empty()) {
+        std::string imported;
+        std::string owning;
+        if (component->isImport()) {
+            imported = "imported ";
+        }
+        if (owningComponent(component) != nullptr) {
+            owning = "' in component '" + owningComponent(component)->name() + "'\n";
+
+        } else {
+            owning = "' in model '" + owningModel(component)->name() + "'\n";
+        }
+        info = "  - " + imported + "component '" + component->name() + owning;
+        addIdMapItem(component->id(), info, idMap);
+    }
+
+    // Variables.
+    for (size_t i = 0; i < component->variableCount(); ++i) {
+        auto item = component->variable(i);
+        if (!item->id().empty()) {
+            info = "  - variable '" + item->name() + "' in component '" + component->name() + "'\n";
+            addIdMapItem(item->id(), info, idMap);
+        }
+        // Equivalent variables.
+        for (size_t e = 0; e < item->equivalentVariableCount(); ++e) {
+            auto equiv = item->equivalentVariable(e);
+            if (equiv != nullptr) {
+                auto equivParent = owningComponent(equiv);
+                if (equivParent != nullptr) {
```

Comment:
@agarny This is where I used them ... now there are checks, but I think it would be cleaner to be able to pass in a nullptr to owningComponent and owningModel and have it return a null rather than a segfault?

Reply:
Sure, but that's besides the point. `owningComponent()` can't be tested with `nullptr` as a parameter. Our tests only have access to the API, not our utilities.

Reply:
I think we don't understand each other.  I would rather not have these two nested if test in my PR version - I'd rather be able to call owningComponent() safely.  This means that if I change the owningComponent function in my PR, and also change my code back to how I wanted it, it will be called during testing, cos that's how I found the segfault in the first place.  It doesn't appear in your PR because your PR doesn't have these functions to check ID.  So coverage will be fine ... ?

Reply:
Oh, ok, got it. Sure, if your code can somehow result in `owningModel()` and `owningComponent()` to be called with `nullptr` as a parameter, then by all means update it in your PR.

(Phew, we are getting there!  Too much noise around me at the office. Am not used to it anymore... and don't want to get used to it again!)